src/Libraries/Microsoft.Extensions.AI.Abstractions/Embeddings/BinaryEmbedding.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Represents an embedding composed of a bit vector.</summary>
+public sealed class BinaryEmbedding : Embedding
+{
+    /// <summary>The embedding vector this embedding represents.</summary>
+    private BitArray _vector;
+
+    /// <summary>Initializes a new instance of the <see cref="BinaryEmbedding"/> class with the embedding vector.</summary>
+    /// <param name="vector">The embedding vector this embedding represents.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="vector"/> is <see langword="null"/>.</exception>
+    public BinaryEmbedding(BitArray vector)
+    {
+        _vector = Throw.IfNull(vector);
+    }
+
+    /// <summary>Gets or sets the embedding vector this embedding represents.</summary>
+    [JsonConverter(typeof(VectorConverter))]
+    public BitArray Vector
+    {
+        get => _vector;
+        set => _vector = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    [JsonIgnore]
+    public override int Dimensions => _vector.Length;
+
+    /// <summary>Provides a <see cref="JsonConverter{BitArray}"/> for serializing <see cref="BitArray"/> instances.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class VectorConverter : JsonConverter<BitArray>
+    {
+        /// <inheritdoc/>
+        public override BitArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            _ = Throw.IfNull(typeToConvert);
+            _ = Throw.IfNull(options);
+
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw new JsonException("Expected string property.");
+            }
+
+            ReadOnlySpan<byte> utf8;
+            byte[]? tmpArray = null;
+            if (reader.HasValueSequence)
+            {
+                ReadOnlySequence<byte> sequence = reader.ValueSequence;
+                long sequenceLength = sequence.Length;
+                if (sequenceLength > int.MaxValue)
+                {
+                    throw new JsonException("Value is too long.");
+                }
+
+                tmpArray = ArrayPool<byte>.Shared.Rent((int)sequenceLength);
+                sequence.CopyTo(tmpArray);
+
+                utf8 = tmpArray.AsSpan(0, (int)sequenceLength);
+            }
+            else
+            {
+                utf8 = reader.ValueSpan;
+            }
+
+            BitArray result = new(utf8.Length);
+
+            for (int i = 0; i < utf8.Length; i++)
+            {
+                result[i] = utf8[i] switch
```

Comment:
This does not account for potential escaping that might exist in the underlying raw string, e.g. the string "01" can be rendered equivalently as `"\u0030\u0031"`. You should consider using the `reader.CopyString` which copies strings to a destination buffer in unescaped form.

Here's an example of how it's being used in one of the built-in STJ converters: https://github.com/dotnet/runtime/blob/0a290e814abaf99f5aaeaacebe76d2e71175db0f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs#L42-L52